### PR TITLE
feat(table): allow for expanding/collapsing all rows in VM selection

### PR DESCRIPTION
**Pull request by seanforyou23**
This PR adds an "expand/collapse" all rows toggle to the select VMs table. It also moves the select all rows function to a bulk select component and refactors the FilterToolbar layout. Should help close https://github.com/konveyor/forklift-ui/issues/36

Screen Shot 2021-06-24 at 4 22 54 PM

### CODE REVIEW

**Comment by mturley** [comment]
@vconzola do you mind giving it a visual review? does the spacing in that toolbar look right?

**Review by vconzola — DISMISSED** [review]
@seanforyou23 @mturley Why are we not using the Bulk selection control (https://www.patternfly.org/v4/guidelines/bulk-selection/) for the VMs table? If we used that along with the Expand/collapse all control (https://www.patternfly.org/v4/components/table/react-demos/#expandcollapse-all) in the toolbar, I think spacing would take care of itself.

**Comment by mturley** [comment]
Hmm I wasn't aware of that control, it seems weird that that exists but there is also a `canSelectAll` prop that we're using to put that checkbox in the header

**Comment by seanforyou23** [comment]
It's a good point @vconzola, I'm not sure why bulk select wasn't used historically but can definitely look into using it here. To @mturley's point, we probably don't want both of these mechanisms to select/deselect all so probably just drop the one from the table header column in favor of the one in the toolbar, right?

I notice the pagination is actually also not housed within the Toolbar itself, so I could take this opportunity to refactor that area of the code so that the bulk expand, bulk select, filter controls, and pagination are all inside the toolbar. This should allow us to drop the Split component usage here and make better use of that `toolbarItems` prop that was recently added for FilterToolbar.

My question is, can we accept this incrementally? If the functionality is ok I'd love to get this bit merged, and then send a follow-up PR that implements bulk select and refactors the FilterToolbar - but no problem if that's not desirable, I can do it all as part of this PR if that's preferred.

**Comment by vconzola** [comment]
@seanforyou23 I'm fine with doing it incrementally, and what you have is OK. But I'd request one change - can you vertically align the filter control with the left side of the checkboxes? It would look a little cleaner. It would also look better if the expand all was in the same row as the select all? I'm a little concerned it's too far away from the table contents to be noticed. But again, this can be an incremental change too.

**Comment by mturley** [comment]
I agree with all of this, we can merge incrementally to keep PRs small but have the end goal of using the Toolbar properly and cleanly for all our tables.

File: src/app/Plans/components/Wizard/SelectVMsForm.tsx
Comment (mturley) [inline src/app/Plans/components/Wizard/SelectVMsForm.tsx:311]: @seanforyou23 there is some out of sync state here since the toggle state is separate from the expanded/collapsed state of the rows. For example, if I click the toggle to expand all the rows, but then I collapse an individual row or all the rows by hand, the toggle is still showing in the expanded position.

IMO we don't need the `expandCollapseToggle` state, and we can instead use the `areAllSelected` (renamed to `areAllExpanded`) property from `useSelectionState` to determine which toggle to show. i.e. if all rows are expanded, show it in the AngleDownIcon mode, otherwise the AngleRightIcon mode. This will make it behave like the indeterminate select-all checkbox: if there are any unselected/unexpanded rows, clicking it selects/expands all, and the deselect/collapse-all action is only available if all are selected/expanded.

Comment (seanforyou23) [reply src/app/Plans/components/Wizard/SelectVMsForm.tsx:313]: Ah ok, I'm starting to see the value in useSelectionState now. Very cool!

**Comment by seanforyou23** [comment]
Ok, I've pushed some updates. I placed the pagination in compact mode in this one place to prevent wrapping of the toolbar items. Is it ok? There may be more I can do in terms of hiding/collapsing toolbar items at given viewport sizes, can explore if needed.

@mturley I got rid of that extra state, but unfortunately it didn't fix the issue of selecting "page" items in the bulk select. Notice the bottom pagination reflects the correct number of selected items, but the checkboxes themselves are not actually checked. In all other scenarios the checkboxes seem to accurately reflect the checked state.. Am I using `setSelectedItems` correctly?

Screen Shot 2021-06-24 at 10 52 27 AM

**Comment by mturley** [comment]
Code LGTM. @vconzola, how does this look visually? I would think we should align the expand/collapse-all caret with the individual row carets, and the bulk select checkbox with the row checkboxes, no?

**Comment by seanforyou23** [comment]
Ah, I see what you mean @mturley - updated as you suggested, looks much better now.

**Comment by mturley** [comment]
@seanforyou23 it looks like there's a border around the expand-all toggle now, was that intentional?

This is looking great btw!

**Comment by seanforyou23** [comment]
Yea, it felt like it was just floating before, so I added the control variant styling. Easier to see the bounds of the selectable area now.

**Review by mturley — APPROVED** [review]
LGTM! Will wait for a final review from @vconzola before merging.

**Pull request merged**
